deepmd/dpmodel/model/linear_model.py

```diff
@@ -0,0 +1,322 @@
+# SPDX-License-Identifier: LGPL-3.0-or-later
+from typing import (
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+
+from deepmd.dpmodel import (
+    FittingOutputDef,
+    OutputVariableDef,
+)
+from deepmd.dpmodel.utils.nlist import (
+    build_multiple_neighbor_list,
+    get_multiple_nlist_key,
+    nlist_distinguish_types,
+)
+
+from .base_atomic_model import (
+    BaseAtomicModel,
+)
+from .dp_atomic_model import (
+    DPAtomicModel,
+)
+from .pair_tab_model import (
+    PairTabModel,
+)
+
+
+class LinearModel(BaseAtomicModel):
+    """Linear model make linear combinations of several existing models.
+
+    Parameters
+    ----------
+    models : list[DPAtomicModel or PairTabModel]
+        A list of models to be combined. PairTabModel must be used together with a DPAtomicModel.
+    weights : list[float] or str
+        If the type is list[float], a list of weights for each model.
+        If "mean", the weights are set to be 1 / len(models).
+        If "sum", the weights are set to be 1.
+        If "zbl", the weights are calculated internally. This only allows the combination of a PairTabModel and a DPAtomicModel.
+    """
+
+    def __init__(
+        self,
+        models: List[Union[DPAtomicModel, PairTabModel]],
+        weights: Union[str, List[float]],
```

Comment:
not a good idea to maintain weight method by strings. 
I would suggest providing an abstract method for defining the weights. and different weight definitions are derived classed implementing the abstract method.

If you want to define the weight by str, the name of the weight for the zbl-interpolated model should not be "zbl", because the weights are nothing to do with the ZBL (which is a pair-wise energy model). You may want to use something like the switch_by_softmin_pair_distance.

Reply:
I thought about defining an interface `Weights`, but that means the user need to create a concrete `ZBLWeights` class before instantiating the `ZBLAtomicModel`. From a developer perspective, that does offer more flexibility, but from a user perspective, that may incur too much extra work.  I will use `str` for now to match the tf implementation.

Reply:
User interacts with deepmd-kit with json script.

The weight is not supposed to be implemented as a class, but a method for the derived method of a base linear model class.

Reply:
I still find the current design strange: the `switch_by_softmin_pair_distance` style suppose that there are only two atomic models, and the order matters. but the constructor does not stop the user provide more than two models.